File: benchmarks/quicksort/bench.py
import time
N = 100_000
LCG_A, LCG_C = 1664525, 1013904223
def partition(arr, lo, hi):
    pivot = arr[hi]; i = lo
    for j in range(lo, hi):
        if arr[j] <= pivot: arr[i], arr[j] = arr[j], arr[i]; i += 1
    arr[i], arr[hi] = arr[hi], arr[i]; return i
def quicksort(arr, n):
    stack = [(0, n-1)]
    while stack:
        lo, hi = stack.pop()
        if lo >= hi: continue
        p = partition(arr, lo, hi)
        if p-1 > lo: stack.append((lo, p-1))
        if p+1 < hi: stack.append((p+1, hi))
rng = 12345
arr = []
for _ in range(N):
    rng = (LCG_A * rng + LCG_C) & 0x7fffffff; arr.append(rng)
t = time.perf_counter()
quicksort(arr, N)
inv = sum(1 for i in range(N-1) if arr[i] > arr[i+1])
print(f"Quicksort {N} ints: {(time.perf_counter()-t)*1000:.1f}ms, inversions={inv} (expect 0)")

File: benchmarks/quicksort/bench.rs
use std::time::Instant;
fn partition(arr: &mut Vec<i32>, lo: usize, hi: usize) -> usize {
    let pivot = arr[hi];
    let mut i = lo;
    for j in lo..hi {
        if arr[j] <= pivot {
            arr.swap(i, j);
            i += 1;
        }
    }
    arr.swap(i, hi);
    i
}
fn quicksort(arr: &mut Vec<i32>, n: usize) {
    let mut stack: Vec<(usize, usize)> = Vec::new();
    stack.push((0, n - 1));
    while let Some((lo, hi)) = stack.pop() {
        if lo < hi {
            let p = partition(arr, lo, hi);
            if p > 0 && p - 1 > lo { stack.push((lo, p - 1)); }
            if p + 1 < hi { stack.push((p + 1, hi)); }
        }
    }
}
fn main() {
    let n = 100_000usize;
    let mut arr: Vec<i32> = Vec::with_capacity(n);
    let mut rng: u32 = 12345;
    for _ in 0..n {
        rng = rng.wrapping_mul(1664525).wrapping_add(1013904223) & 0x7fffffff;
        arr.push(rng as i32);
    }
    let t = Instant::now();
    quicksort(&mut arr, n);
    let elapsed = t.elapsed().as_millis();
    let inv = arr.windows(2).filter(|w| w[0] > w[1]).count();
    println!("Quicksort {} ints: {}ms, inversions={} (expect 0)", n, elapsed, inv);
}
